ENH: inconsistent naming convention for read_excel column selection (#4988)

abarber4gh opened the pull request:
- update API to use ‘usecols’ instead of ‘parse_cols’.  still functionally the same as ‘parse_col’,
- added test cases for ‘usecols’, added assert_produces_warning(FutureWarning) to other test
cases that use ‘parse_cols’.
- refactor column use column parsing to only occur once per sheet.
- updated whats new with deprecated parse_col argument.

 - [X] closes #4988 
 - [X] tests added / passed
 - [X] passes ``git diff upstream/master --name-only -- '*.py' | flake8 --diff``
 - [X] whatsnew entry
 - [X] packers.* BENCHMARKS NOT SIGNIFICANTLY CHANGED.

abarber4gh committed:
ENH: inconsistent naming convention for read_csv and read_excel column selection (#4988)

update API to use ‘usecols’ instead of ‘parse_cols’.  still functionally the same as ‘parse_col’,
added test cases for ‘usecols’, added assert_produces_warning(FutureWarning) to other test
cases that use ‘parse_cols’.

refactor column use column parsing to only occur once per sheet.

updated whats new with deprecated parse_col argument and other enhancements
to usecols functionality.
update documentation to show new usecols functionality in read_excel().

abarber4gh committed:
ENH: inconsistent naming convention for read_csv and read_excel column selection (#4988)

update API to use ‘usecols’ instead of ‘parse_cols’.  still functionally the same as ‘parse_col’,
added test cases for ‘usecols’, added assert_produces_warning(FutureWarning) to other test
cases that use ‘parse_cols’.

refactor column use column parsing to only occur once per sheet.

updated whats new with deprecated parse_col argument and other enhancements
to usecols functionality.
update documentation to show new usecols functionality in read_excel().

chris-b1 commented:
FYI, just discussing a bug related to usecols #15316, might be easiest to work that change into your refactoring, or can be separate, cc @stanleyguan

jreback commented on doc/source/whatsnew/v0.21.0.txt:
```diff
@@ -35,6 +35,7 @@ Other Enhancements
 - ``RangeIndex.append`` now returns a ``RangeIndex`` object when possible (:issue:`16212`)
 - :func:`to_pickle` has gained a protocol parameter (:issue:`16252`). By default, this parameter is set to `HIGHEST_PROTOCOL <https://docs.python.org/3/library/pickle.html#data-stream-format>`__
 - :func:`api.types.infer_dtype` now infers decimals. (:issue: `15690`)
+- :func:`read_excel` now allows a column character list (E.G. ['A', 'C', 'D']) with the ``usecols`` parameter (:issue:`4988`).
```
we are listed this in deprecations so this is unecessary

jreback commented on doc/source/whatsnew/v0.21.0.txt:
```diff
@@ -60,6 +61,7 @@ Other API Changes
 Deprecations
 ~~~~~~~~~~~~
 - :func:`read_excel()` has deprecated ``sheetname`` in favor of ``sheet_name`` for consistency with to_excel() (:issue:`10559`).
+- :func:`read_excel()` has deprecated ``parse_cols`` in favor of ``usecols`` for consistency with other read_ functions (:issue:`4988`).
```
``pd.read_*`` functions

jreback commented on pandas/io/excel.py:
```diff
@@ -111,12 +111,16 @@
     * If string then indicates comma separated list of column names and
       column ranges (e.g. "A:E" or "A,C,E:F")
+parse_cols : int or list, default None
+    .. deprecated:: 0.21.0
```
hmm, didn't know sphinx had this deprecated tag. @TomAugspurger @jorisvandenbossche should maybe change other deprecations to use this if it looks nice.

abarber4gh replied:
appears as:
> **parse_cols** _: int or list, default None_
> _Deprecated since version 0.21.0:_ Use _usecols_ instead

TomAugspurger replied:
http://pandas-docs.github.io/pandas-docs-travis/generated/pandas.read_excel.html Has an example from sheetname. Looks ok

jreback replied:
@TomAugspurger if we like this (and lgtm), let's make an issue to updated all DEPRECATED in codebase with this directive

jreback commented on pandas/tests/io/test_excel.py:
```diff
@@ -169,54 +169,155 @@ def test_parse_cols_int(self):
+        dfref = self.get_csv_refdf('test1')
+        dfref = dfref.reindex(columns=['A', 'B', 'C'])
+        df1 = self.get_exceldf('test1', 'Sheet1', index_col=0, usecols=3)
+        df2 = self.get_exceldf('test1', 'Sheet2', skiprows=[1], index_col=0,
```
so you need to change ALL tests to use the new one (usecols), except for a single test to actually hit the deprecation.

jreback commented on pandas/tests/io/test_excel.py:
```diff
@@ -169,54 +169,155 @@ def test_parse_cols_int(self):
+
+        tm.assert_frame_equal(df1, dfref, check_names=False)
+        tm.assert_frame_equal(df2, dfref, check_names=False)  # backward compat
+
```
are these new tests?

abarber4gh replied:
yes, new tests for the new functionality.

abarber4gh committed:
implement changes request in PR#16488
- removed usecols mention in Other Enhancments section,
remains in Deprecations.
- removed test_parse_* test methods in favor of test_usecols_* methods.
- changed parse_cols to usecols in test_read_one_empty_col_* instead
of catching warning.

abarber4gh committed:
implement changes request in PR#16488
- removed usecols mention in Other Enhancments section,
remains in Deprecations.
- removed test_parse_* test methods in favor of test_usecols_* methods.
- changed parse_cols to usecols in test_read_one_empty_col_* instead
of catching warning.

abarber4gh committed:
no message

abarber4gh committed:
no message

jreback commented on pandas/io/excel.py:
```diff
@@ -312,11 +327,6 @@ def _range2cols(areas):
             """
-            def _excel2num(x):
-                "Convert Excel column name like 'AB' to 0-based column index"
-                return reduce(lambda s, a: s * 26 + ord(a) - ord('A') + 1,
```
is there a reason you are changing this code? does this involve the deprecation somehow? if you are cleaning/fixing, pls do in another PR.

abarber4gh replied:
code changed to allow passing list of strings, similar to other `read_*` functions, and keep the `read_excel` "superpower" to pass a string that is parsed as mentioned in the original issue (#4988, point 3).

jreback replied:
and that's fine, but needs to be in a separate PR from the deprecation change.

abarber4gh replied:
ok, opened #16510 for this functionality.  will re-submit with only the argument change (parse_cols -> usecols).

jreback commented on pandas/tests/io/test_excel.py:
```diff
@@ -165,58 +165,100 @@ def setup_method(self, method):
         tm.assert_frame_equal(df2, dfref, check_names=False)
+        tm.assert_frame_equal(df3, dfref, check_names=False)  # backward compat
 
-    def test_parse_cols_str(self):
```
leave the original tests structure (sure you can change the name to conform), but don't change the tests (in THIS PR).

abarber4gh replied:
tests are back to original but with changed function & kwarg names.

abarber4gh committed:
change tests keyword from parse_cols to usecol.

abarber4gh committed:
change tests keyword from parse_cols to usecol.

abarber4gh committed:
change parse_cols to usecols

abarber4gh committed:
change parse_cols to usecols

abarber4gh committed:
removed excess blank line.

abarber4gh committed:
removed excess blank line.

TomAugspurger commented:
@abarber4gh can you rebase, now that the excel tests are running?

abarber4gh committed:
add `deprecate_kwarg` from `_decorators`
add `check_stacklevel=False` to `test_excel_oldindex_format()`

abarber4gh committed:
add `deprecate_kwarg` from `_decorators`
add `check_stacklevel=False` to `test_excel_oldindex_format()`

abarber4gh committed:
TST: ujson tests are not being run (#16499) (#16500)

closes #16499

gfyoung committed:
DOC: Remove preference for pytest paradigm in assert_raises_regex (#16518)

neirbowj committed:
TST: Specify HTML file encoding on PY3 (#16526)

TomAugspurger committed:
BUG: Fixed tput output on windows (#16496)

keitakurita committed:
BUG: Incorrect handling of rolling.cov with offset window (#16244)

TomAugspurger committed:
TST: Avoid global state in matplotlib tests (#16539)

Replaces most uses of implicit global state from matplotlib in
test_datetimelike.py. This was potentially causing random failures
where a figure expected to be on a new, blank figure would instead
plot on an existing axes (that's the guess at least).

VincentLa committed:
DOC: Update to docstring of DataFrame(dtype) (#14764) (#16487)

* Adding some more documentation on dataframe with regards to dtype

* Making example for creating dataframe from np matrix easier

ProsperousHeart committed:
DOC: correct docstring examples (#3439) (#16432)

jtratner committed:
Fix unbound local with bad engine (#16511)

Tafkas committed:
return empty MultiIndex for symmetrical difference on equal MultiIndexes (#16486)

JosephWagner committed:
BUG: select_as_multiple doesn't respect start/stop kwargs GH16209 (#16317)

jreback committed:
BUG: Bug in .resample() and .groupby() when aggregating on integers (#16549)

closes #16361

mattip committed:
COMPAT: cython str-to-int can raise a ValueError on non-CPython (#16563)

pepicello committed:
CLN: raise correct error for Panel sort_values (#16532)

TomAugspurger committed:
BUG: Fixed pd.unique on array of tuples (#16543)

pvomelveny committed:
BUG: Allow non-callable attributes in aggregate function. Fixes GH16405 (#16458)

TomAugspurger committed:
Strictly monotonic (#16555)

gfyoung committed:
COMPAT: Consider Python 2.x tarfiles file-like (#16533)

CRP committed:
BUG: Fixed to_html ignoring index_names parameter

closes #16493

Author: Christian Prinoth <christian@prinoth.name>
Author: Tom Augspurger <tom.w.augspurger@gmail.com>
Author: Christian Prinoth <c.prinoth@CPmbpro.local>
Author: Jeff Reback <jeff.reback@twosigma.com>

This patch had conflicts when merged, resolved by
Committer: Jeff Reback <jeff.reback@twosigma.com>

Closes #16495 from CRP/bugfix_16493 and squashes the following commits:

567ae69 [Jeff Reback] doc corrections
8429f9a [Tom Augspurger] Fixed lint error
469a0e6 [Christian Prinoth] BUG: fix for bug 16493
20d512f [Christian Prinoth] BUG: fix for bug 16493
6bef829 [Christian Prinoth] BUG: fix for bug 16493
426565e [Christian Prinoth] BUG: fix for bug 16493
a40820d [Christian Prinoth] BUG: fix for bug 16493

economy committed:
BUG: fixed wrong order of ordered labels in pd.cut()

closes #16459

Author: economy <the.economy@gmail.com>

This patch had conflicts when merged, resolved by
Committer: Jeff Reback <jeff.reback@twosigma.com>

Closes #16466 from economy/fix_cut and squashes the following commits:

29128b3 [economy] comments and whatsnew edits
3898b72 [economy] BUG: fixed wrong order of ordered labels in pd.cut()

jreback committed:
fix linting

Commit:
TST: writing invalid table names to sqlite (#16464)

* Add test for bug #13206.

* Improve test by reading back the values from sql and comparing. Also fixes coding style violation.

neirbowj committed:
TST: Skip test_database_uri_string if pg8000 importable (#16528)

sullivanrs committed:
DOC: Remove incorrect elements of PeriodIndex docstring (#16553)

* DOC: Remove incorrect elements of PeriodIndex docstring

See #9056.

* Removed trailing space

TomAugspurger committed:
TST: Make HDF5 fspath write test robust (#16575)

The test_write_fspath_all test would fail on the HDF5 example
occasionally (about 1/100 in my experience). Apparently you don't get an
identical HDF5 every single time. This refactors that test out to its own where
we write and read both versions, and compare equality that way.

dsm054 committed:
ENH: add .ngroup() method to groupby objects (#14026) (#14026)

OlegShteynbuk committed:
make null lowercase a missing value (#16534)

gfyoung committed:
MAINT: Drop has_index_names input from read_excel (#16522)

rhendric committed:
BUG: reimplement MultiIndex.remove_unused_levels (#16565)

chrisgorgo committed:
Adding 'n/a' to list of strings denoting missing values (#16079)

TomAugspurger committed:
API: Make is_strictly_monotonic_* private (#16576)

jorisvandenbossche committed:
DOC: change doc build to python 3.6 (#16545)

* DOC: change doc build to python 3.6

* Remove pinning of pyqt to 4.x

* Remove pinning of openpyxl

* Add xsel to doc build for clipboard

jreback committed:
DOC: whatsnew 0.20.2 edits (#16587)

funnycrab committed:
DOC: Fix typo in timeseries.rst (#16590)

TomAugspurger committed:
PERF: vectorize _interp_limit (#16592)

* PERF: vectorize _interp_limit

* CLN: remove old implementation

* fixup! CLN: remove old implementation

benjello committed:
DOC: Fix typo in merge doc for validate kwarg (#16595)

makmanalp committed:
BUG: convert numpy strings in index names in HDF #13492 (#16444)

* BUG: Handle numpy strings in index names in HDF5 #13492

* REF: refactor to _ensure_str

bpraggastis committed:
ERRR: Raise error in usecols when column doesn't exist but length matches (#16460)

* gh-14671 Check if usecols with type string contains a subset of names, if not throws an error

* tests added for gh-14671, expected behavior of simultaneous use of usecols and names unclear so these tests are commented out

* Review comments

TomAugspurger committed:
DOC: Whatsnew fixups (#16596)

TomAugspurger committed:
DOC: Update release.rst

jreback committed:
BUG: pickle compat with UTC tz's (#16611)

closes #16608

jhelie committed:
Fix some lgtm alerts (#16613)

jreback committed:
BLD: fix numpy on 3.6 build as 1.13 was released but no deps are built for it (#16633)

dsm054 committed:
BUG: Fix Series.get failure on missing NaN (#8569) (#16619)

dsm054 committed:
TST: NaN in MultiIndex should not become a string (#7031) (#16625)

dsm054 committed:
TST: verify we can add and subtract from indices (#8142) (#16629)

preddy5 committed:
BUG: conversion of Series to Categorical (#16557)

fix #16524

jreback committed:
BLD: fix numpy on 2.7 build as 1.13 was released but no deps are built for it (#16633) (#16650)

tswast committed:
CLN: make license file machine readable (#16649)

Splits extra information about the license and copyright holders to
AUTHORS.md.

jreback committed:
fix pytest-xidst version as 1.17 appears buggy (#16652)

jreback committed:
COMPAT: numpy 1.13 test compat (#16654)

* COMPAT: numpy 1.13 test compat

* CI: fix doc build to 1.12

abarber4gh committed:
Merge branch 'issue#4988' of https://github.com/abarber4gh/pandas into issue#4988

* 'issue#4988' of https://github.com/abarber4gh/pandas:
  add `deprecate_kwarg` from `_decorators` add `check_stacklevel=False` to `test_excel_oldindex_format()`
  removed excess blank line.
  change parse_cols to usecols
  change tests keyword from parse_cols to usecol.
  no message
  ENH: inconsistent naming convention for read_csv and read_excel column selection (#4988)
  implement changes request in PR#16488 - removed usecols mention in Other Enhancments section, remains in Deprecations. -      removed test_parse_* test methods in favor of test_usecols_* methods. - changed parse_cols to usecols in test_read_one_empty_col_* instead of catching warning.
rebased

abarber4gh committed:
rebase with #16522 changes.

jreback commented:
can you rebase / update according to comments

jreback commented:
closing as stale. pls ping to reopen if you want to continue.
